cdc/sink/mysql.go

```diff
@@ -357,60 +356,121 @@ func newMySQLSink(sinkURI *url.URL, dsn *dmysql.Config, filter *util.Filter, opt
-		jobs <- dmls
-	}
-	close(jobs)
+	return concurrentExec(ctx, rowGroups, s.params.workerCount, s.params.maxTxnRow, s.execDMLs)
+}
 
-	nWorkers := s.params.workerCount
+func concurrentExec(ctx context.Context, rowGroups map[string][]*model.RowChangedEvent, nWorkers, maxTxnRow int, execDMLs func(context.Context, []*model.RowChangedEvent) error) error {
 	if nWorkers == 0 {
 		nWorkers = defaultParams.workerCount
 	}
 
-	if len(rowGroups) < nWorkers {
-		nWorkers = len(rowGroups)
-	}
-	eg, _ := errgroup.WithContext(ctx)
+	errCh := make(chan error, 1)
+	rowChs := make([]chan *model.RowChangedEvent, 0, nWorkers)
+	var jobWg sync.WaitGroup
+
 	for i := 0; i < nWorkers; i++ {
-		eg.Go(func() error {
-			for rows := range jobs {
-				err := rowLimitIterator(rows, s.params.maxTxnRow,
-					func(rows []*model.RowChangedEvent) error {
-						return errors.Trace(s.execDMLs(ctx, rows))
-					})
-				if err != nil {
-					return errors.Trace(err)
+		rowCh := make(chan *model.RowChangedEvent, 1024)
+		worker := mysqlSinkWorker{
+			rowCh:     rowCh,
+			errCh:     errCh,
+			jobWg:     &jobWg,
+			maxTxnRow: maxTxnRow,
+			execDMLs:  execDMLs,
+		}
+		go worker.run(ctx)
+		rowChs = append(rowChs, rowCh)
+	}
+	causality := newCausality()
+	rowChIdx := 0
+	sendFn := func(row *model.RowChangedEvent, keys []string, idx int) {
+		causality.add(keys, idx)
+		jobWg.Add(1)
+		rowChs[idx] <- row
+	}
+	for groupKey, rows := range rowGroups {
+		for _, row := range rows {
+			keys := row.Keys
+			if len(keys) == 0 {
+				keys = []string{groupKey}
+			}
+			if conflict, idx := causality.detectConflict(keys); conflict {
+				if idx >= 0 {
+					sendFn(row, keys, idx)
+					continue
 				}
+				jobWg.Wait()
+				causality.reset()
 			}
-			return nil
-		})
+			idx := rowChIdx % len(rowChs)
+			sendFn(row, keys, idx)
+			rowChIdx++
+		}
+	}
+	for i := range rowChs {
+		close(rowChs[i])
+	}
+	jobWg.Wait()
+	select {
+	case err := <-errCh:
+		return err
+	default:
+		return nil
 	}
-	return eg.Wait()
 }
 
-func rowLimitIterator(rows []*model.RowChangedEvent, maxTxnRow int, fn func([]*model.RowChangedEvent) error) error {
-	start := 0
-	end := maxTxnRow
-	for end < len(rows) {
-		lastTs := rows[end-1].Ts
-		for ; end < len(rows); end++ {
-			if lastTs < rows[end].Ts {
-				break
-			}
+type mysqlSinkWorker struct {
+	rowCh     chan *model.RowChangedEvent
+	errCh     chan error
+	jobWg     *sync.WaitGroup
+	maxTxnRow int
+	execDMLs  func(context.Context, []*model.RowChangedEvent) error
+}
+
+func (w *mysqlSinkWorker) run(ctx context.Context) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Error("mysql sink worker panic",
+				zap.Reflect("r", r), zap.Stack("stack trace"))
 		}
-		if err := fn(rows[start:end]); err != nil {
-			return errors.Trace(err)
+	}()
+	var rows []*model.RowChangedEvent
+	for row := range w.rowCh {
+		rows = append(rows, row)
+		rows = w.fetchAllPendingEvent(rows)
+		// TODO: Add retry
+		err := w.execDMLs(ctx, rows)
+		w.trySendErr(err)
```

Comment:
should we check err is nil here? or maybe the channel will be filled with a nil error but real error is not sent to

Reply:
Great catch.